module/zfs/zfs_vnops.c

```diff
@@ -123,7 +123,7 @@ zfs_holey_common(znode_t *zp, ulong_t cmd, loff_t *off)
 
 	/* Flush any mmap()'d data to disk */
 	if (zn_has_cached_data(zp, 0, file_sz - 1))
-		zn_flush_cached_data(zp, B_FALSE);
+		zn_flush_cached_data(zp, B_TRUE);
```

Comment:
This is interesting.  On the [CentOS 9](https://build.openzfs.org/builders/CentOS%209%20x86_64%20%28TEST%29/builds/2338/steps/shell_4/logs/console) we hit the following list corruption in this callpath.  I don't recall seeing this before.

```
[ 6712.352678] (node->next == LIST_POISON1) is equivalent to (node->prev == LIST_POISON2)
[ 6712.352680] PANIC at list.h:187:list_link_active()
[ 6712.352682] Showing stack for process 1102662
[ 6712.352684] CPU: 1 PID: 1102662 Comm: seekflood Tainted: P           OE     -------  ---  5.14.0-362.24.1.el9_3.x86_64 #1
[ 6712.352686] Hardware name: Amazon EC2 m5d.large/, BIOS 1.0 10/16/2017
[ 6712.352687] Call Trace:
[ 6712.352689]  <TASK>
[ 6712.352692]  dump_stack_lvl+0x34/0x48
[ 6712.352699]  spl_panic+0xd1/0xe9 [spl]
[ 6712.352722]  ? list_head+0x9/0x30 [zfs]
[ 6712.353003]  ? dbuf_rele_and_unlock+0x135/0x520 [zfs]
[ 6712.353247]  ? RW_READ_HELD+0x15/0x30 [zfs]
[ 6712.353492]  ? dnode_hold_impl+0x8e6/0x10c0 [zfs]
[ 6712.353737]  spl_assert+0x17/0x20 [zfs]
[ 6712.353993]  list_link_active+0x53/0x70 [zfs]
[ 6712.354251]  dnode_is_dirty+0x5d/0x190 [zfs]
[ 6712.354497]  dmu_offset_next+0x67/0x210 [zfs]
[ 6712.354732]  zfs_holey_common+0x94/0x190 [zfs]
[ 6712.354990]  zfs_holey+0x4b/0x80 [zfs]
[ 6712.355242]  zpl_llseek+0x89/0xc0 [zfs]
[ 6712.355482]  ksys_lseek+0x81/0xc0
[ 6712.355487]  do_syscall_64+0x5c/0x90
[ 6712.355491]  ? do_syscall_64+0x69/0x90
[ 6712.355492]  entry_SYSCALL_64_after_hwframe+0x72/0xdc
```

Reply:
Hmm, quite interesting. I've seen this too, but the context is a bit different...

So I've written a tool that distributes ZTS tests across many worker VMs running on my personal dev machine. Each of these workers run a clone of a base image with tmpfs as a backing store. In this setup, the workers tend to PANIC in various places every so many runs. I've chalked these problems up to the test environment as it's full of contention, so the tool has logic to reboot and retry on PANIC as a workaround.

This particular PANIC appears in test run logs 7 times, and each of them is exactly like the CentOS 9 output showing a crash in seekflood, for example:

```
[ 3130.540354] EXT4-fs (vdb): unmounting filesystem e57f42fe-9cab-43eb-a304-998dfad58f79.
[ 3130.669342] EXT4-fs (vdb): mounted filesystem 3be99c1e-f9d6-4d46-836f-8877510dd3d0 r/w without journal. Quota mode: none.
[ 3131.275029] loop0: detected capacity change from 0 to 8388608
[ 3131.312348] loop1: detected capacity change from 0 to 8388608
[ 3131.350089] loop2: detected capacity change from 0 to 8388608
[ 3269.583387] (node->next == LIST_POISON1) is equivalent to (node->prev == LIST_POISON2)
[ 3269.583393] PANIC at list.h:187:list_link_active()
[ 3269.583432] Showing stack for process 492785
[ 3269.583446] CPU: 2 PID: 492785 Comm: seekflood Tainted: P           OE      6.6.8-200.fc39.x86_64 #1
[ 3269.583449] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.16.2-1.fc37 04/01/2014
[ 3269.583451] Call Trace:
[ 3269.583455]  <TASK>
[ 3269.583462]  dump_stack_lvl+0x47/0x60
[ 3269.583492]  spl_panic+0x100/0x120 [spl]
[ 3269.583512]  ? dnode_hold_impl+0xb1c/0x1550 [zfs]
[ 3269.583723]  multilist_link_active+0x63/0x80 [zfs]
[ 3269.583914]  dnode_is_dirty+0x69/0x1a0 [zfs]
[ 3269.584098]  dmu_offset_next+0x7e/0x270 [zfs]
[ 3269.584281]  zfs_holey+0xe3/0x270 [zfs]
[ 3269.584459]  zpl_llseek+0xa5/0x150 [zfs]
[ 3269.584632]  ksys_lseek+0x6c/0xb0
[ 3269.584650]  do_syscall_64+0x60/0x90
[ 3269.584658]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584665]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584667]  ? syscall_exit_to_user_mode+0x2b/0x40
[ 3269.584672]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584674]  ? do_syscall_64+0x6c/0x90
[ 3269.584677]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584679]  ? syscall_exit_to_user_mode+0x2b/0x40
[ 3269.584681]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584685]  ? do_syscall_64+0x6c/0x90
[ 3269.584687]  ? srso_alias_return_thunk+0x5/0x7f
[ 3269.584689]  ? do_syscall_64+0x6c/0x90
[ 3269.584692]  ? do_syscall_64+0x6c/0x90
[ 3269.584694]  ? do_syscall_64+0x6c/0x90
[ 3269.584698]  entry_SYSCALL_64_after_hwframe+0x6e/0xd8
[ 3269.584707] RIP: 0033:0x7fdd16b4a6db
[ 3269.584723] Code: 5b 07 0d 00 64 c7 00 0d 00 00 00 eb b0 e8 cd d8 01 00 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 00 f3 0f 1e fa b8 08 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 05 c3 0f 1f 40 00 48 8b 15 21 07 0d 00 f7 d8
[ 3269.584725] RSP: 002b:00007ffc28f057a8 EFLAGS: 00000246 ORIG_RAX: 0000000000000008
[ 3269.584729] RAX: ffffffffffffffda RBX: 0000000000000003 RCX: 00007fdd16b4a6db
[ 3269.584733] RDX: 0000000000000003 RSI: 0000000000000000 RDI: 0000000000000003
[ 3269.584734] RBP: 00007ffc28f05850 R08: 0000000000000064 R09: 0000000000000000
[ 3269.584738] R10: 0000000000001000 R11: 0000000000000246 R12: 0000000000000444
[ 3269.584739] R13: 0000000000000003 R14: 0000000000000223 R15: 0000000000000000
[ 3269.584744]  </TASK>
```

Unfortunately I don't have the code build logged, but this run is from 2024-03-04. This is before I was aware of #15933 at all. (I was probably testing changes to `dnode_next_offset` towards making `llseek` not need to flush txgs...) Which is to say it predates this patch, but maybe the other patch causes it?

Anyway since this is raising your eyebrows I'll try a bit harder to reproduce it and get a root cause.

p.s. I intend to debug some of the other PANICs at some point, but I'm not really sure what to make of most of them just yet.

Reply:
Actually nevermind most of that. I remember now, this is `dnode_is_dirty` still having racy reads of both `dn_dirty_link` and `dn_dirty_records` vs. the sync code clearing them both while writing.

The _dirtying_ code holds locks, but the sync side does _not_ since most sync code assumes txg state is quiescent. So it's possible to observe half of the list inactive and fail the assertion.

My intended fix to here is to stop using dnode-level dirty detection. The correct way IMO is to use `db_dirtycnt` on each dbuf while walking the block tree in `dnode_next_offset` to stop at dirty data. So then we don't need to have any special handling.

Reply:
To confirm the above, I ran `tests/functional/cp_files/cp_stress` 50 times at c28f94f32e (head) and 50 times at 590ee7a13c (this PR).

* At head, the test failed 6 out of 50 times with `PANIC at list.h:187:list_link_active()`.
* At this PR, the test failed 5 out of 50 times with the same error.

```
[   39.799977] (node->next == LIST_POISON1) is equivalent to (node->prev == LIST_POISON2)
[   39.799982] PANIC at list.h:187:list_link_active()
[   39.800025] Showing stack for process 1528
[   39.800031] CPU: 3 PID: 1528 Comm: seekflood Tainted: P           OE      6.6.8-200.fc39.x86_64 #1
[   39.800035] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 1.16.2-1.fc37 04/01/2014
[   39.800036] Call Trace:
[   39.800042]  <TASK>
[   39.800048]  dump_stack_lvl+0x47/0x60
[   39.800075]  spl_panic+0x100/0x120 [spl]
[   39.800091]  ? srso_alias_return_thunk+0x5/0x7f
[   39.800097]  ? dnode_hold_impl+0x8a3/0x1070 [zfs]
[   39.800363]  spl_assert+0x17/0x20 [zfs]
[   39.800624]  list_link_active+0x53/0x70 [zfs]
[   39.800886]  dnode_is_dirty+0x61/0x190 [zfs]
[   39.801028]  dmu_offset_next+0x72/0x220 [zfs]
[   39.801159]  zfs_holey_common+0x94/0x180 [zfs]
[   39.801299]  zfs_holey+0x4f/0x80 [zfs]
[   39.801437]  zpl_llseek+0x8c/0xd0 [zfs]
[   39.801573]  ksys_lseek+0x6c/0xb0
[   39.801588]  do_syscall_64+0x60/0x90
```

I'm convinced this is just the pre-existing data race.

Reply:
With the caveat that I'm in transit right now and have barely thought about this, I'm inclined to agree. If we force the the mapped page to flush, but the dnode is still dirty, then there's nothing to cover the gap while the dnode is off-list.

(which was the intent of #15615, and I've been intending to reply to your comment there but never quite got around to it, sigh).

Reply:
> I'm convinced this is just the pre-existing data race.

Agreed, that looks like it's the case to me.  Thanks for looking in to this, I'll get this fix  merged.